Soft-pin the default versions of Aquarius, Brizo, Keeper Contracts &  Pleuston. Add new --stable option

ttmc opened the pull request:
A recurring problem for new users is that the `--latest` and stable (default) sets of versions (of Aquarius, Brizo, Keeper Contracts and Pleuston) don't work with whatever they are trying to use (typically something based on squid-py or squid-js).

This pull request introduces a new `--squid-py-v0.4.2-compatible` option which pins:

```text
export AQUARIUS_VERSION=v0.1.8
export BRIZO_VERSION=v0.2.0
export KEEPER_VERSION=v0.6.12
```

which is based on [Marcus's Manta Ray Release Matrix](https://docs.google.com/spreadsheets/d/1hiB-85LIboRtc0WokeL53D754MjG7guQBwyBjeVqSk0/edit?usp=sharing) for Mantaray 0.4.0.

It doesn't pin PLEUSTON_VERSION because squid-py users don't usually run Pleuston anyway.

More version-set-pinning options like that could be added in the future. This one is just a starting point to test the idea.

Also:
- updated the `README.md` file to give new users some commands which actually work, i.e. using the new `--squid-py-v0.4.2-compatible` option. I also incorporated Jerry's technique of pruning all the Docker things before calling `start_ocean.sh`. (There is a note that that is overkill and that they could add the options `--purge` and `--force-pull` instead.)
- clarified that they no longer have to put Azure credentials into `brizo.env` (unless they want to use Azure Storage).
- clarified how the `--latest` and default (stable) version sets work, including what happens when you set one of the environment variables like `KEEPER_VERSION`.

ttmc commented:
The way `--latest` works today is that it gets the Docker images of Aquarius, Brizo, Keeper Conracts, and Pleuston with the Docker image tag `latest` (built from the `develop` branch of those repos). Unfortunately, that means that the _meaning_ of `latest` changes over time and that can cause things to break. Changing `--latest` to mean some specific/fixed versions is an interesting idea, but those specific/fixed versions would have to be changed from time to time, and that would cause problems again.

You can always override `--latest` for a specific component (e.g. Aquarius) by setting a specific environment variable (e.g. `AQUARIUS_VERSION=v0.1.8`.). I'm reluctant to change how `latest` works because we're all used to that by now, and it's used all over the place (e.g. when testing in Travis).

Another way would be to tell the user to do their own pinning, i.e. by setting the values of:

- AQUARIUS_VERSION
- BRIZO_VERSION
- KEEPER_VERSION
- PLEUSTON_VERSION

in some preceding bash commands or maybe even in the call to `start_ocean.sh` itself, e.g.

```bash
export AQUARIUS_VERSION=v0.1.8
export BRIZO_VERSION=v0.2.0
export KEEPER_VERSION=v0.6.12
./start_ocean.sh --no-pleuston --local-spree-node
```

That's not very clear or pretty, but it's explicit and works. We're already doing that to some extent: we often pin KEEPER_VERSION for Travis tests, and Marcus has some pet bash scripts which already do the above. I could add a table of some common version sets to the Barge README.md file (and get rid of the new `--squid-py-v0.4.2-compatible` option.

MarcusJones commented:
Great initiative @ttmc - I vote having a new tag `--stable` which fixes the versions to the following; (there have been a few updates)
```
export AQUARIUS_VERSION=v0.1.10
export BRIZO_VERSION=v0.2.2
export KEEPER_VERSION=v0.6.12
```

And we can keep `--latest` as-is.

ttmc commented:
> I vote having a new tag `--stable` which fixes the versions to the following...

The issue with that is that:

- `--latest` means use the `latest` Docker image tags for the Aquarius, Brizo, Keeper Contracts and Pleuston containes
- (default, no option) means use the `stable` Docker image tags for...

so one would _think_ (by analogy) that `--stable` would mean to use the `stable` Docker image tags. Maybe we could use some other option name like `--recent`?

I'm leaning towards giving a table (in the README.md file) with some recent sets of versions that work together (like in the Manta Ray Release Matrix). Let them pick the set that suits them.

kremalicious commented:
this seems to just get more complicated with more flags. The default, without any flag, should be stable and fixed versions. `--latest` gets unstable, latest development versions. Everything else breaks most user's expectations towards what those flags mean. The problem isn't that we don't have enough flags, it's that there is no stable matrix. So the default fixed versions in this script should be the ones where the full flow of publishing and consuming works. If it doesn't we don't have anything stable.

As for this PR's `--squid-py-v0.4.2-compatible` flag I second Aitor, users can export the specific versions of the components. So the matrix table route in readme sounds best to me, with a code snippet how to export those versions on command line when running `start_ocean.sh`

Also, if we add more of those flags we should start actually adding proper options, so e.g. instead of `--local-spree-node` it should be `--node spree` or something.

ttmc commented:
> The default, without any flag, should be stable and fixed versions.

I like that idea, assuming you mean Aquarius, Brizo, Keeper Contracts and Pleuston would all be frozen at specific versions set inside the `start_ocean.sh` script. That set could change over time and it would be listed in the table in the Barge `README.md` file.

This would mean changing the meaning of the default but it _would_ be nice if the default always worked.

ttmc commented:
Based on the discussion so far, I pushed some commits to change the approach:

- Removed the `--squid-py-v0.4.2-compatible` option.
- Added a new `--stable` option to run all the `stable` Docker image tags of Aquarius, Brizo, Keeper Contracts and Pleuston.
- Created a new set of default values for `AQUARIUS_VERSION`, `BRIZO_VERSION`, `KEEPER_VERSION` and `PLEUSTON_VERSION` (but these will be overridden if they are already set before running `start_ocean.sh`). I used the versions suggested by Marcus above, and `PLEUSTON_VERSION=latest`. I know that Pleuston & squid-js aren't working with that set of versions right now. I added a note about that in the README.md file. Hopefully we'll have a fully-working set of versions soon.
- Updated the `README.md` file:
  - Tell new users to run a local Spree node (local network), and to _not_ run Pleuston, for now (because it doesn't work fully with the other default components right now).
  - Added a new table to show exactly what the `--latest`, `--stable` and default options mean for the component Docker image tags.
  - Added a concrete example of how to override the Docker image tag used by a particular component.
  - In the existing table of options, moved the two Docker-related options one after the other.

ttmc commented:
Please hold off on merging this pull request until we have a good set of default versions that all work together.

ttmc commented:
The latest commit (https://github.com/oceanprotocol/barge/pull/119/commits/98a285bc26577f792cea0956ad9ac4184e87dd1c) was to resolve a merge conflict that got introduced when the default network got changed from Nile to Spree (which arose from merging pull request #134).

aaitor commented on start_ocean.sh:
```diff
@@ -12,6 +12,12 @@ export BRIZO_ENV_FILE="${DIR}/brizo.env"
 COMPOSE_DIR="${DIR}/compose-files"
 
+# Default versions of Aquarius, Brizo, Keeper Contracts and Pleuston
+export AQUARIUS_VERSION=${AQUARIUS_VERSION:-v0.1.10}
```
These versions are not right, see: https://github.com/oceanprotocol/barge/issues/127#issuecomment-479445626

ttmc replied:
Yes I was waiting for there to be a stable set of versions working together. Now that one exists, I will update those version numbers.
